caledonia/src/prehashed.rs

```diff
@@ -0,0 +1,331 @@
+}
+impl Setup {
+    /// Setup algorithm taking a Params as input and returning setup parameters (u,d,q)
+    pub fn new(params: &Params) -> Self {
+        let e = E;
+        let log_2 = |x: f64| x.log2();
+        let loge = log_2(e);
+        let logloge = log_2(loge);
+        let log3 = log_2(3.0);
+
+        let n_p_f64 = params.n_p as f64;
+        let n_f_f64 = params.n_f as f64;
+        let lognpnf = log_2(n_p_f64 / n_f_f64);
+        let lambda_sec = params.lambda_sec as f64;
+        let lambda_rel = params.lambda_rel as f64;
+
+        // We define the parameters according to Section 3.2, Corrollary 2
+        let u_f64 = (lambda_sec + log_2(lambda_rel + log3) + 1.0 - logloge) / lognpnf;
+        let u = u_f64.ceil() as usize;
+        let d_f64 = 16.0 * u_f64 * (lambda_rel + log3) / loge;
+        let d = d_f64.ceil() as usize;
+        let q = (2.0 * (lambda_rel + log3) / (d_f64 * loge)).recip().ceil() as usize;
+
+        let check = ((d_f64 * d_f64 * loge) / (9.0 * (lambda_rel + log3))).ceil() as usize;
+        assert!(params.n_p >= check);
+
+        Setup {
+            n_p: params.n_p,
+            u,
+            d,
+            q,
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
+    t: usize,
+    // Round candidate tuple
+    s_list: Vec<Data>,
+    /// Round candidate hash
+    h: Vec<u8>,
+    /// Round candidate hash mapped to [1, n_p]
+    h_usize: usize,
+    /// Approximate size of set Sp to lower bound
+    n_p: usize,
+}
+
+impl Round {
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
+    /// We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h1(data: Vec<Vec<u8>>, n_p: usize) -> (Hash, usize) {
+        let digest = utils::combine_hashes(data);
+        return (digest, utils::oracle(&digest, n_p));
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(t: usize, n_p: usize) -> Round {
+        let data = [t.to_ne_bytes().to_vec()].to_vec();
+        let (h, h_usize) = Round::h1(data, n_p);
+        Round {
+            t,
+            s_list: Vec::new(),
+            h: h.to_vec(),
+            h_usize,
+            n_p,
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: Data) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = Vec::new();
+        data.push(r.h.clone());
+        data.push(s.to_vec());
+        let (h, h_usize) = Round::h1(data, r.n_p);
+        Round {
+            t: r.t,
+            s_list,
+            h: h.to_vec(),
+            h_usize,
+            n_p: r.n_p,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
```

Comment:
These classes exist in `bounded.rs` too and they share the same interface, so we can use Traits to extract the common functionality